include/af/memory.h

```diff
@@ -0,0 +1,339 @@
+
+AFAPI af_err af_buffer_info_get_ptr(void** ptr, af_buffer_info buf);
+
+AFAPI af_err af_buffer_info_get_event(af_event* event, af_buffer_info buf);
+
+AFAPI af_err af_buffer_info_set_ptr(af_buffer_info buf, void* ptr);
+
+AFAPI af_err af_buffer_info_set_event(af_buffer_info buf, af_event event);
+
+/// \brief Disassociates the \ref af_event from the \ref af_buffer_info object
+///
+/// Gets the \ref af_event and disassociated it from the af_buffer_info object.
+/// Deleting the af_buffer_info object will not affect this event.
+///
+/// param[out] event The \ref af_event that will be disassociated. If NULL no
+/// event is
+///                   returned and the event is NOT freed
+/// param[in] buf The target \ref af_buffer_info object
+/// \returns AF_SUCCESS
+AFAPI af_err af_unlock_buffer_info_event(af_event* event, af_buffer_info buf);
+
+/// \brief Disassociates the pointer from the \ref af_buffer_info object
+///
+/// Gets the pointer and disassociated it from the \ref af_buffer_info object.
+/// Deleting the af_buffer_info object will not affect this pointer.
+///
+/// param[out] event The \ref pointer that will be disassociated. If NULL no
+/// pointer is
+///                  returned and the data is NOT freed.
+/// param[in] buf The target \ref af_buffer_info object
+/// \returns AF_SUCCESS
+AFAPI af_err af_unlock_buffer_info_ptr(void** ptr, af_buffer_info buf);
+
+////////////////////////////////////////////////////////////////////////////////
+// Memory Manager API
+////////////////////////////////////////////////////////////////////////////////
+
+typedef void (*af_memory_manager_initialize_fn)(af_memory_manager);
+
+typedef void (*af_memory_manager_shutdown_fn)(af_memory_manager);
+
+typedef af_buffer_info (*af_memory_manager_alloc_fn)(af_memory_manager, size_t,
+                                                     /* bool */ int);
+
+typedef size_t (*af_memory_manager_allocated_fn)(af_memory_manager, void*);
+
+typedef void (*af_memory_manager_unlock_fn)(af_memory_manager, void*, af_event,
+                                            /* bool */ int);
+
+typedef void (*af_memory_manager_garbage_collect_fn)(af_memory_manager);
+
+typedef void (*af_memory_manager_print_info_fn)(af_memory_manager, char*, int);
+
+typedef void (*af_memory_manager_usage_info_fn)(af_memory_manager, size_t*,
+                                                size_t*, size_t*, size_t*);
+
+typedef void (*af_memory_manager_user_lock_fn)(af_memory_manager, void*);
+
+typedef void (*af_memory_manager_user_unlock_fn)(af_memory_manager, void*);
+
+typedef int (*af_memory_manager_is_user_locked_fn)(af_memory_manager, void*);
+
+typedef size_t (*af_memory_manager_get_mem_step_size_fn)(af_memory_manager);
+
+typedef size_t (*af_memory_manager_get_max_bytes_fn)(af_memory_manager);
+
+typedef unsigned (*af_memory_manager_get_max_buffers_fn)(af_memory_manager);
+
+typedef void (*af_memory_manager_set_mem_step_size_fn)(af_memory_manager,
```

Comment:
Problem is we have public API functions in `af/device.h` iirc that call these functions - things like `af::setStepSize` or `af::deviceGC`. We need to the default memory manager and user-defined memory manager to share a custom interface, and I think we concluded there wasn't really a to do that without making these functions part of the public API, but I'm open to suggestions.

One idea would be to have default implementations in the base class that throw for `garbage_collect` and `[get/set]_mem_step_size` as being not implemented - that way, it would be possible for the user to not define function pointers for those things in their custom implementation.

Reply:
They should call default derived memory manager directly with these functions in case they are called by the user. Those functions are directly tied to that memory manager anyway so it makes sense. It looks like we create an instance of default memory manager even if the user will override it with the custom memory manager.  It would not return correct information wrt. to the active memory manager but I don't see why those functions would be called if a custom memory manager is used. 

I think `deviceGC` should be part of the interface because ArrayFire will call that function if an allocation fails. max buffer and step size related functions should not be exposed.

Reply:
@umar456 — that's not correct — once the user sets a memory manager, the default memory manager is destroyed. There is no "partial control" of memory state by the default memory manager — everything is handed over to the user-defined memory manager.

This goes back to [the discussion we had before](https://github.com/arrayfire/arrayfire/pull/2461#issuecomment-475482975) about why this has to be the case. Because all of the methods in the current public API are called by AF internals, we *must* implement all of those methods in some way. Having some "default" implementation of those methods ultimately breaks because then there are two sources of truth: the internal and the external memory managers. There are lots of examples with how this breaks:
- Say the user doesn't implement `usageInfo`. This is called by the JIT/essential for AF internals. If there's fallback to a "default" implementation, that implementation *must* know about the state of the user-defined memory manager in order to give correct output. There's no way to do that without inspecting the user's implementation directly... which is exactly what having a `usageInfo` function on the user-defined memory manager would accomplish.
- `checkMemoryLimit` has the same issue. It's called by the JIT, and an internal implementation that doesn't know about externally-defined allocator state can't actually know if we're at the memory limit
- [this applies to all of the user-defined functions in the external API - otherwise, I would've gotten rid of them]

Reply:
I see what you mean about destroying the default memory manager but I am not sure if I am explaining myself well. I also agree that the user should update all functions. I am suggesting we us a smaller set because some of the functions being exposed are not necessary. If there is a API call that calls those functions they should return an error or an invalid value. Internally we should try to change the current behavior to avoid calling those functions because they are too dependent on the internals of the default memory manager and they may not map well to a custom memory manager.

I should explain how we use those functions inside the library. Functions like `deviceMemoryInfo` and `checkMemoryLimit` are there to find reasonable heuristics based on the memory pressure. For example the deviceMemoryInfo function is just used to shortcut a slightly more expensive check further down in the function. The information returned from 'deviceMemoryInfo' is not necessary for JIT to function properly. 'checkMemoryLimit' does almost the exact same function as the deviceMemoryInfo call in the JIT but its only used in the CPU backend. Its used to avoid scenarios where too many operations are queued on the system and their combined memory allocation is taking too much memory.

In the past we had a good understanding of the internals of the memory manager so we could directly use the counts to perform those heuristics. With custom memory managers these heuristics will be useless and harmful because the custom MM will have different performance characteristics.

A better approach would be to replace these functions with a getMemoryPressure function which returns a floating point value between 0 and 1. 0 will indicate no memory pressure and 1 will indicate the memory subsystem is nearly full. We can replace the existing function that compares the value returned by a certain value(user defined). For the default implementation you can set it so that the function returns zero and 1 if:

'lock_bytes > getMaxBytes() || lock_buffers > getMaxBuffers()'

This will give you the same functionality as the current implementation. Then where ever those functions are used replace with 'getMemoryPressure() < 0.75f)'. or what ever value the user finds appropriate. We can find a good middle later with a few experiments.

If the user calls these functions then these functions should return an error message because a custom memory manager is used. I don't see a reason to have those functions if the user is going to change them anyway.

I hope this makes my thoughts more clear. I really appreciate this conversation and if there is anything that is unclear please let me know. There is a lot of hidden details here that may not be familiar to you.

Reply:
[added]